File: demo_v2_threshold/demo_v2/backend/app/services/metric_validation.py
"""Validate metric values against definition rules."""

from decimal import Decimal, InvalidOperation

from app.models.metric_definition import MetricDefinition


class MetricValidationError(Exception):
    def __init__(self, code: str, message: str) -> None:
        self.code = code
        self.message = message
        super().__init__(message)


def validate_metric_value(definition: MetricDefinition, raw_value: Decimal | float | int | str) -> Decimal:
    try:
        value = Decimal(str(raw_value))
    except (InvalidOperation, ValueError) as exc:
        raise MetricValidationError(definition.code, "Value must be numeric") from exc

    rules = definition.validation_rules or {}
    data_type = definition.data_type

    if data_type == "integer":
        if value != value.to_integral_value():
            raise MetricValidationError(definition.code, "Value must be an integer")
        value = Decimal(int(value))

    if "min" in rules and value < Decimal(str(rules["min"])):
        raise MetricValidationError(definition.code, f"Value must be >= {rules['min']}")
    if "max" in rules and value > Decimal(str(rules["max"])):
        raise MetricValidationError(definition.code, f"Value must be <= {rules['max']}")
    if rules.get("min_exclusive") is True and "min" in rules and value <= Decimal(str(rules["min"])):
        raise MetricValidationError(definition.code, f"Value must be > {rules['min']}")

    return value

File: demo_v2_threshold/demo_v2/frontend/src/utils/metricValidation.ts
import type { MetricDefinition } from "../types/metrics";

export function validateMetricInput(
  definition: MetricDefinition,
  raw: string,
): string | null {
  const trimmed = raw.trim();
  if (trimmed === "") {
    return "Value is required";
  }

  const num = Number(trimmed);
  if (Number.isNaN(num)) {
    return "Must be a number";
  }

  const rules = definition.validation_rules ?? {};

  if (definition.data_type === "integer" && !Number.isInteger(num)) {
    return "Must be a whole number";
  }

  if (rules.min !== undefined && num < Number(rules.min)) {
    return `Must be at least ${rules.min}`;
  }
  if (rules.max !== undefined && num > Number(rules.max)) {
    return `Must be at most ${rules.max}`;
  }
  if (rules.min_exclusive && rules.min !== undefined && num <= Number(rules.min)) {
    return `Must be greater than ${rules.min}`;
  }

  return null;
}

export function inputModeForDataType(
  dataType: string,
): "numeric" | "decimal" | undefined {
  if (dataType === "integer") {
    return "numeric";
  }
  if (dataType === "decimal") {
    return "decimal";
  }
  return undefined;
}
